hibernate-core/src/main/java/org/hibernate/query/Pager.java

```diff
@@ -0,0 +1,247 @@
+/*
+ * Hibernate, Relational Persistence for Idiomatic Java
+ *
+ * License: GNU Lesser General Public License (LGPL), version 2.1 or later
+ * See the lgpl.txt file in the root directory or http://www.gnu.org/licenses/lgpl-2.1.html
+ */
+package org.hibernate.query;
+
+import org.hibernate.Incubating;
+
+import java.util.Iterator;
+import java.util.List;
+import java.util.NoSuchElementException;
+import java.util.function.Consumer;
+import java.util.stream.Stream;
+import java.util.stream.StreamSupport;
+
+import static java.util.Collections.emptyList;
+import static java.util.Spliterator.IMMUTABLE;
+import static java.util.Spliterator.NONNULL;
+import static java.util.Spliterator.ORDERED;
+import static java.util.Spliterators.spliteratorUnknownSize;
+
+/**
+ * Allows bidirectional iteration over pages of query results, useful for processing
+ * query results in batches.
+ * <pre>
+ * Pager&lt;Book&gt; bookPager =
+ *		 session.createSelectionQuery("from Book where title like :title", Book.class)
+ *				.setParameter("title", titlePattern)
+ *				.getResultPager(Page.first(pageSize));
+ * while (bookPager.hasResults()) {
+ *	 List&lt;Book&gt; books = bookPager.getResultList();
+ *	 ...
+ *	 session.clear();
+ *	 bookPager.next();
+ * }
+ * </pre>
+ * Or simply:
+ * <pre>
+ * session.createSelectionQuery("from Book where title like :title", Book.class)
+ *		.setParameter("title", titlePattern)
+ *		.getResultPager(Page.first(pageSize))
+ *		.forEachRemainingPage(books -> {
+ *			...
+ *			session.clear();
+ *		});
+ * </pre>
+ *
+ * @param <R> the query result type
+ *
+ * @see Page
+ * @see SelectionQuery#getResultPager(Page)
+ *
+ * @author Gavin King
+ * @since 6.3
+ */
+@Incubating
+public class Pager<R> {
+	private final SelectionQuery<R> query;
+	private List<R> results;
+	private Page page;
+	private boolean hasMoreResults;
+	private int lastPage = -1;
+
+	Pager(SelectionQuery<R> query, Page initialPage) {
+		this.query = query;
+		this.page = initialPage;
+		this.hasMoreResults = true;
+		execute();
+	}
+
+	/**
+	 * @return the {@link SelectionQuery} backing this pager.
+	 */
+	public SelectionQuery<R> getQuery() {
+		return query;
+	}
+
+	/**
+	 * @return the current {@linkplain Page page}.
+	 */
+	public Page getCurrentPage() {
+		return page;
+	}
+
+	/**
+	 * @return the current page of results, as a list.
+	 */
+	public List<R> getResultList() {
+		return results;
+	}
+
+	/**
+	 * Advance to the next page of results.
+	 *
+	 * @see Page#next()
+	 */
+	public void next() {
```

Comment:
The name `next()` might give the feeling that the Pager is some kind of Iterator while it is a bit different.
Maybe it could be something like 'executeNext()' to also convey the meaning that an SQL query will be executed when this is called?

Or maybe the Pager _could_ be some kind of Iterator and `next()` could return a Page and the results could be in the Page?